docs/content/Configuration.md
# hichipper parameters 

Running
```
hichipper --help
```
shows the parameters that can be used in this software package as reproduced below.

```
Usage: hichipper [OPTIONS] MANIFEST

  A preprocessing and QC pipeline for HiChIP data.

Options:
  --out TEXT                    Output directory name; must not be an already existing directory [Required]
  --min-dist TEXT               Minimum distance; default = 5000
  --max-dist TEXT               Peak padding width (applied on both left and
                                right); default = 2000000
  --macs2-string TEXT           String of arguments to pass to MACS2; only is
                                called when peaks are set to be called;
                                default = "-q 0.01 --extsize 147 --nomodel"
  --macs2-genome TEXT           Argument to pass to the -g variable in MACS2
                                (mm for mouse genome; hs for human genome);
                                default = "hs"
  --peak-pad TEXT               Peak padding width (applied on both left and
                                right); default = 500
  --merge-gap TEXT              Merge nearby peaks (after all padding is
                                complete); default = 500
  --keep-temp-files             Keep temporary files?
  --skip-background-correction  Skip restriction fragment aware background
                                correction?
  --skip-resfrag-pad            Skip restriction fragment aware padding
  --skip-qc                     Skip QC report generation?
  --skip-diffloop               Skip analyses in diffloop (e.g. Mango loop
                                calling; .rds generation)
  --make-ucsc                   Make additional output files that can support
                                viewing in UCSC genome browser; requires tabix
                                and htslib tools.
  --keep-samples TEXT           Comma separated list of sample names to keep;
                                ALL (special string) by default
  --ignore-samples TEXT         Comma separated list of sample names to
                                ignore; NONE (special string) by default
  --read-length TEXT            Length of reads from sequencing runs; default = 75
  --version                     Show the version and exit.
  --help                        Show this message and exit.
```
 
Running
```
hichipper --version
```  
will show the version of this package currently installed. 

```
hichipper, version 0.6.0
```
Check the badge up top to see if a newer version is available or try directly through `pip`:

```
pip install hichipper --upgrade
```

Unless these flags are supplied, the pipeline will attempt to run. Minimally sufficient parameters include
the `--out` flag and a `.yaml` file as shown in the example executions. Below are some explanations of the
additional parameters than can be configured when executing the pipeline. 

## Restriction-fragment aware padding
<img src="https://github.com/aryeelab/hichipper/blob/master/docs/content/media/RE_PAD.png?raw=True" style="width:100%;"><br>

## Parameter explanations

Most of the parameter options are fairly straight forward. Running `hichipper --version` or `hichipper --help`
doesn't run the tool but supplies the information noted above. Otherwise, the default run mode requires 
a `.yaml` file supplied in addition to the `--out` parameter, which specifies the output directory of the run. 
Users can decide to customize final output by using boolean flags or supply variable text input. The following 
cartoon shows a graphical overview of important parameters to consider when running **hichipper**.

<img src="https://github.com/aryeelab/hichipper/blob/master/docs/content/media/parameters.png?raw=True" style="width:100%;"><br>

As noted in orange, defined peaks are automatically padded by some integer width from the `--peak-pad` flag. By default, 
this pad extends 500 base pairs in either direction. Padding the peaks boosts the number of PETs that can be mapped to loops. 
For example, `PET II` would not be considered in loop since the left end of the read does not overlap with the called peak (black).
However, it does overlap with the padded peak, so it is retained with padding. When two peaks are close to one another, they may
be merged using the `--merge-gap` command. As suggested in the figure, the padded peaks `B` and `C` may be sufficiently close to be merged into a single anchor. 
Note that this can lead to some PETs becoming self-ligation (e.g. `I-III`). Note, the `--merge-gap` command is equivalent to running 
[bedtools merge -d](http://bedtools.readthedocs.io/en/latest/content/tools/merge.html) on the padded anchors.
By default, the `merge-gap` is 500 base pairs. Specifying this to `0` can cause issues, particularly when the width of a `PET` spans multiple peaks.

The `dist` or distance between two peaks is noted in black as the center of two peaks. The `--min-dist` flag is the smallest
and `--max-dist` is the largest integer number that ensures this distance falls between to be considered in a loop. These defaults
are 5Kb and 2Mb as smaller reads are likely self-ligations whereas larger reads are unlikely to be biologically real loops.

From our inspection of the HiChIP data, we determined that reads for putative loops localize to the edges of restriciton fragments and that
increasing the padding to the edges of fragments can prove beneficial for maximizing the read density associated with loops.
By default, **hichipper** adds additional padding to the edges of restriction fragments unless the `--skip-resfrag-pad` pad is thrown.
To give an example, consider peak `C`. Under the default options, the anchor corresponding to this peak would span restriction
fragments `7` and `9` (assuming that the `merge-gap` is small enough such that `B` and `C` are not merged). This is because the
additional restriction fragment padding occurs after the original padding (orange), and **hichipper** padds to the edges of whatever
fragment pads are overlapping the peak. However, if the user specified `--peak-pad 0`, the anchor corresponding to `C` would only span
restriction motifs `8` and `9` since the peak only sits on that singular fragment. While we have found the default options to be
sensible, the user can adjust these padding parameters to potentially increase precision of anchors calls possibly at the expense of PET density or vice-versa.  

Finally, we note the `--macs2-genome` and `--macs2-string` which by default has parameters that we felt were suited appropriately for processing HiChIP data.
However, users can modify these when performing peak-calling from HiChIP data directly. 

## User parameter recommendations
- If `R` is not in the system or if the `R` package dependencies could not be installed, the following flags should be added:
```
--skip-resfrag-pad --skip-diffloop --skip-qc --skip-background-correction
```
- In the current version of **hichipper**, the novel background correction implementation is quite memory intense. Thus, _users running **hichipper** on a laptop
or other low RAM machine_ should likely skip the adaptive background correction. 
```
--skip-background-correction
```

<br><br>


NEWS
# 0.6.0
- Fixed inconsistencies with duplicates

# 0.3.0
- New qcReport in .html format

# 0.2.7
- Fixed qcReport for single sample

# 0.2.6
- Addressed issue with variable read length

# 0.2.5
- Fixed bug with single sample
- Read length no longer important (need a new param though...)
- Grey out loops that aren't significant in DNAlandscapeR
- Improved QC reporting

# 0.2.4
- Added diffloop output
- Better flow control/informative error messages
- Added a few QC metrics

# 0.2.2
- Fixed bug in QC report generation
- Full helper script integration

# 0.2.1
- Addition of MANIFEST file to add helper scripts

# 0.2.0
- Initial addition to PyPi

# 0.1.0
- Added NEWS


hichipper/cli.py
import click
import os
import os.path
import sys
import shutil
import yaml
import shutil
import random
import string
import time
from pkg_resources import get_distribution
from subprocess import call, check_call

def parse_manifest(manifest):
    samples = []
    if manifest.endswith(('.yaml', '.yml')):
        with open(manifest, 'r') as f: 
            m = yaml.load(f)
        return m
    else:
        click.echo(gettime() + "Please specify a valid .yaml file for samples")

#Grab Sample Names
def get_subdirectories(dir):
    return [name for name in os.listdir(dir)
            if os.path.isdir(os.path.join(dir, name))]

@click.command()
@click.option('--out', default="hichipper_out", required=True, help='Output directory name; must not be already existing [Required]')
@click.option('--min-dist', default="5000", help='Minimum distance ; default = 5000')
@click.option('--max-dist', default="2000000", help='Peak padding width (applied on both left and right); default = 2000000')
@click.option('--macs2-string', default="-q 0.01 --extsize 147 --nomodel", help='String of arguments to pass to MACS2; only is called when peaks are set to be called; default = "-q 0.01 --extsize 147 --nomodel"')
@click.option('--macs2-genome', default="hs", help='Argument to pass to the -g variable in MACS2 (mm for mouse genome; hs for human genome); default = "hs"')
@click.option('--peak-pad', default="500", help='Peak padding width (applied on both left and right); default = 500')
@click.option('--merge-gap', default="500", help='Merge nearby peaks (after all padding is complete; default = 500')
@click.option('--keep-temp-files', is_flag=True, help='Keep temporary files?')
@click.option('--skip-background-correction', is_flag=True, help='Skip restriction fragment aware background correction?')
@click.option('--skip-resfrag-pad', is_flag=True, help='Skip restriction fragment aware padding')
@click.option('--skip-qc', is_flag=True, help='Skip QC report generation?')
@click.option('--skip-diffloop', is_flag=True, help='Skip analyses in diffloop (e.g. Mango loop calling; .rds generation)')
@click.option('--make-ucsc', is_flag=True, help='Make additional output files that can support viewing in UCSC genome browser; requires tabix and htslib tools.')
@click.option('--keep-samples', default="ALL", help='Comma separated list of sample names to keep; ALL (special string) by default')
@click.option('--ignore-samples', default="NONE", help='Comma separated list of sample names to ignore; NONE (special string) by default')
@click.option('--read-length', default="75", help='Length of reads from sequencing runs; default = 75')
@click.argument('manifest')
@click.version_option()

def main(manifest, out, min_dist, max_dist, macs2_string, macs2_genome, peak_pad, merge_gap, keep_temp_files, skip_background_correction, skip_resfrag_pad, skip_qc, skip_diffloop, read_length, keep_samples, ignore_samples, make_ucsc):
	"""A preprocessing and QC pipeline for HiChIP data."""
	__version__ = get_distribution('hichipper').version
	if os.path.exists(out):
		sys.exit("ERROR: Output path (%s) already exists." % out)
	
	def gettime(): # Matches `date` in Linux
		return(time.strftime("%a ") + time.strftime("%b ") + time.strftime("%d ") + time.strftime("%X ") + time.strftime("%Z ") + time.strftime("%Y")+ ": ")
	
	os.mkdir(out)
	logf = open(out + "/" + out + ".hichipper.log", 'w')
	click.echo(gettime() + "Starting hichipper pipeline v%s" % __version__, file = logf)
	click.echo(gettime() + "Starting hichipper pipeline v%s" % __version__)
	click.echo(gettime() + "Parsing user parameters")
	script_dir = os.path.dirname(os.path.realpath(__file__))
	outfolder = os.path.abspath(out)  
	cwd = os.getcwd()
	click.echo(gettime() + "Executed from: %s" % cwd, logf)
	click.echo(gettime() + "Output folder: %s" % outfolder, logf) 
	m = parse_manifest(manifest)
	click.echo(gettime() + "Parsed manifest as follows: ", logf)
	click.echo(m, logf)
	
	peaks = m['peaks'][0]
	resfrags = m['resfrags'][0]
	hicprooutput = m['hicpro_output'][0]
	
	# Determine if restriction fragment calling is happening or not
	if not (skip_resfrag_pad or skip_background_correction):
		if not os.path.isfile(resfrags):
			sys.exit('ERROR: Could not find the restriction fragment file ' + resfrags + '; either correctly specify file in .yaml or use the --skip-resfrag-pad and --skip-background-correction flags')
		
	# Determine samples from user-defined input
	bwt_samples = os.popen('ls ' + hicprooutput + '/bowtie_results/bwt2').read().strip().split("\n")
	hic_samples = os.popen('ls ' + hicprooutput + '/hic_results/data').read().strip().split("\n")
	
	def intersect(a, b):
		return list(set(a) & set(b))

	samples = intersect(bwt_samples, hic_samples)
	
	if(keep_samples != "ALL"):
		keeplist = keep_samples.split(",")
		click.echo(gettime() + "Intersecting detected samples with user-retained ones: " + keep_samples)
		samples = intersect(samples, keeplist)
		
	if(ignore_samples != "NONE"):
		igslist = ignore_samples.split(",")
		for byesample in igslist:
			click.echo(gettime() + "Attempting to remove " + byesample + " from processing", logf)
			if byesample in samples: samples.remove(byesample)
	
	if not len(samples) > 0:
		sys.exit('ERROR: Could not import any samples from the user specification; check flags, logs and .yaml configuration; quitting')
	
	halfLength = int(float(read_length)/2)
	if not halfLength> 0:
		sys.exit('ERROR: Specify appropriate read length > 1')
	
	
	# Make sure that files are present
	def verify_sample(sample, boo):
		bt = os.popen('ls ' + hicprooutput + '/bowtie_results/bwt2/' + sample + "/*.pairstat").read().strip().split("\n")
		hc1 = os.popen('ls ' + hicprooutput + '/hic_results/data/' + sample + "/*.RSstat").read().strip().split("\n")
		hc2 = os.popen('ls ' + hicprooutput + '/hic_results/data/' + sample + "/*.*Pairs").read().strip().split("\n")
		hc3 = os.popen('ls ' + hicprooutput + '/hic_results/data/' + sample + "/*_allValidPairs").read().strip().split("\n")
		
		if(boo):
			return((len(bt) > 0) and (len(hc1) > 0) and (len(hc2) % 5) == 0 and (len(hc3) > 0))
	
	for sample in samples:
		if not verify_sample(sample, True):
			sys.exit('ERROR: Missing hic_results or bowtie_results files for ' + sample + "; either exclude the sample or manage the file architecture/input")
	
	click.echo(gettime() + "Determined that the following samples are good to go: ", logf)
	click.echo(samples, logf)

	# Set up peaks files if necessary; moreover, specify vector of peaks / sample
	peakfilespersample = []
	click.echo(gettime() + "User defined peaks specification: " + peaks, logf)
	
	# Call peaks integrating over all samples for self ligation reads
	macs2Error = 'ERROR: macs2 peak calling was not successful; make sure macs2 is in the environment and that *Pairs files exist for samples. Finally, note that only the narrowPeak file is being considered here, which may be a problem if you tried to do broad peak calling.'
	if(peaks == "COMBINED,SELF"):
		click.echo(gettime() + "Calling one set of peaks from HiChIP self-ligation reads across all samples.", logf)
		for sample in samples:
		    os.system('cat ' + hicprooutput + '/hic_results/data/' + sample + "/*DEPairs >> " + out + "/allDEandSCpairs.Pairs.tmp")
		    os.system('cat ' + hicprooutput + '/hic_results/data/' + sample + "/*SCPairs >> " + out + "/allDEandSCpairs.Pairs.tmp")
		os.system('''awk -v RL=''' + str(halfLength)+r''' '{print $2"\t"$3-RL"\t"$3+RL}' '''+ out + '''/allDEandSCpairs.Pairs.tmp | awk '$2 > 0 {print $0}' > ''' + out + '''/allDEandSCpairs.bed.tmp''')
		os.system('''awk -v RL=''' + str(halfLength)+r''' '{print $5"\t"$6-RL"\t"$6+RL}' '''+ out + '''/allDEandSCpairs.Pairs.tmp | awk '$2 > 0 {print $0}' >> ''' + out + '''/allDEandSCpairs.bed.tmp''')
		macs2cmd = 'macs2 callpeak -t ' + out + '/allDEandSCpairs.bed.tmp ' + macs2_string + " -g " + macs2_genome + " -B -f BED --verbose 0 -n " + out + '/allSamples_temporary'
		click.echo(gettime() +  "macs2 command: " + macs2cmd, logf)
		os.system(macs2cmd)
		if not os.path.isfile(out + '/allSamples_temporary_peaks.narrowPeak'):
			sys.exit(macs2Error)
		
		# Run hichipper modified background correction for all samples combined using self ligation reads
		if (not skip_background_correction):
			click.echo(gettime() + "Performing hichipper-modified background peak calling")
			call(['Rscript', os.path.join(script_dir, 'lambdaProcess.R'), resfrags,  os.path.join(out, "allSamples_temporary_treat_pileup.bdg"), os.path.join(out,"allSamples_temporary_control_lambda.bdg"), out])
			click.echo(gettime() + "Modified background signal; performing peak calling")
			click.echo(gettime() +  "MACS2 TEXT OUTPUT INCOMING")
			os.system("macs2 bdgcmp -t " + os.path.join(out,"adjustedTreatment.bdg.tmp") + " -c " + os.path.join(out, "adjustedBackground.bdg.tmp") + " -m qpois -o " + os.path.join(out,"hichipper_qvalue.bdg.tmp"))
			os.system("macs2 bdgpeakcall -i " + os.path.join(out,"hichipper_qvalue.bdg.tmp") + " -c 2 -l 147 -g 100 -o " + os.path.join(out,"hichipper_peaks.bed"))
			os.system("mv " + os.path.join(out,"hichipper_peaks.bed") + " " + os.path.join(out,"allSamples_temporary_hichipperPeaks.bed"))
			peakfilespersample = [out + '/allSamples_temporary_hichipperPeaks.bed'] * len(samples)
		else:
			peakfilespersample = [out + '/allSamples_temporary_peaks.narrowPeak'] * len(samples)
	
	# Call peaks from within each sample using self-ligation reads
	elif(peaks == "EACH,SELF"):
		click.echo(gettime() + "Calling peaks from self-ligation reads for each sample", logf)
		for sample in samples:
		    os.system('cat ' + hicprooutput + '/hic_results/data/' + sample + "/*DEPairs >> " + out + "/" + sample +"DEandSCpairs.Pairs.tmp")
		    os.system('cat ' + hicprooutput + '/hic_results/data/' + sample + "/*SCPairs >> " + out + "/" + sample +"DEandSCpairs.Pairs.tmp")
		    os.system('''awk -v RL=''' + str(halfLength)+r''' '{print $2"\t"$3-RL"\t"$3+RL}' '''+ out + '''/''' + sample +'''DEandSCpairs.Pairs.tmp | awk '$2 > 0 {print $0}' >  ''' + out + '''/''' + sample +'''DEandSCpairs.bed.tmp''')
		    os.system('''awk -v RL=''' + str(halfLength)+r''' '{print $5"\t"$6-RL"\t"$6+RL}' '''+ out + '''/''' + sample +'''DEandSCpairs.Pairs.tmp | awk '$2 > 0 {print $0}' >> ''' + out + '''/''' + sample +'''DEandSCpairs.bed.tmp''')
		    macs2cmd = 'macs2 callpeak -t ' + out + '/' + sample + 'DEandSCpairs.bed.tmp ' + macs2_string + " -g " + macs2_genome + " -B -f BED --verbose 0 -n " + out + '/' + sample + "_temporary"
		    click.echo(gettime() +  "macs2 command: " + macs2cmd, logf)
		    os.system(macs2cmd)
		    if not os.path.isfile(out + '/' + sample + '_temporary_peaks.narrowPeak'):
				sys.exit(macs2Error)
			
		    # Run hichipper modified background correction for each sample using self-ligation reads
		    if (not skip_background_correction):
				click.echo(gettime() + "Performing hichipper-modified background peak calling")
				os.system('Rscript ' + os.path.join(script_dir, 'lambdaProcess.R') + " " + resfrags + " " + out + '/' + sample + "_temporary_treat_pileup.bdg " + out + '/' + sample + "_temporary_control_lambda.bdg " + out)
				click.echo(gettime() + "Modified background signal; performing peak calling")
				click.echo(gettime() +  "MACS2 TEXT OUTPUT INCOMING")
				os.system("macs2 bdgcmp -t " + os.path.join(out,"adjustedTreatment.bdg.tmp") + " -c " + os.path.join(out, "adjustedBackground.bdg.tmp") + " -m qpois -o " + os.path.join(out,"hichipper_qvalue.bdg.tmp"))
				os.system("macs2 bdgpeakcall -i " + os.path.join(out,"hichipper_qvalue.bdg.tmp") + " -c 2 -l 147 -g 100 -o " + os.path.join(out,"hichipper_peaks.bed"))
				os.system("mv " + os.path.join(out,"hichipper_peaks.bed") + " " + out + "/" + sample + "_temporary_hichipperPeaks.bed")
				peakfilespersample.append(out + '/' + sample + '_temporary_hichipperPeaks.bed')
		    else:
				peakfilespersample.append(out + '/' + sample + '_temporary_peaks.narrowPeak')
		    
	
	# Call peaks integrating over all samples using  all reads
	elif(peaks == "COMBINED,ALL"):
		click.echo(gettime() + "Calling one set of peaks from all HiChIP reads across all samples.", logf)
		for sample in samples:
		    os.system('cat ' + hicprooutput + '/hic_results/data/' + sample + "/*Pairs >> " + out + "/all.Pairs.tmp")
		os.system('''awk -v RL=''' + str(halfLength)+r''' '{print $2"\t"$3-RL"\t"$3+RL}' '''+ out + '''/all.Pairs.tmp | awk '$2 > 0 {print $0}' > ''' + out + '''/allpairs.bed.tmp''')
		os.system('''awk -v RL=''' + str(halfLength)+r''' '{print $5"\t"$6-RL"\t"$6+RL}' '''+ out + '''/all.Pairs.tmp | awk '$2 > 0 {print $0}' >> ''' + out + '''/allpairs.bed.tmp''')
		macs2cmd = 'macs2 callpeak -t ' + out + '/allpairs.bed.tmp ' + macs2_string + " -g " + macs2_genome + " -B -f BED --verbose 0 -n " + out + '/allSamples_temporary'
		click.echo(gettime() +  "macs2 command: " + macs2cmd, logf)
		os.system(macs2cmd)
		if not os.path.isfile(out + '/allSamples_temporary_peaks.narrowPeak'):
			sys.exit(macs2Error)
			
		# Run hichipper modified background correction for all samples combined using all
		if (not skip_background_correction):
			click.echo(gettime() + "Performing hichipper-modified background peak calling")
			call(['Rscript', os.path.join(script_dir, 'lambdaProcess.R'), resfrags,  os.path.join(out, "allSamples_temporary_treat_pileup.bdg"), os.path.join(out,"allSamples_temporary_control_lambda.bdg"), out])
			click.echo(gettime() + "Modified background signal; performing peak calling")
			click.echo(gettime() +  "MACS2 TEXT OUTPUT INCOMING")
			os.system("macs2 bdgcmp -t " + os.path.join(out,"adjustedTreatment.bdg.tmp") + " -c " + os.path.join(out, "adjustedBackground.bdg.tmp") + " -m qpois -o " + os.path.join(out,"hichipper_qvalue.bdg.tmp"))
			os.system("macs2 bdgpeakcall -i " + os.path.join(out,"hichipper_qvalue.bdg.tmp") + " -c 2 -l 147 -g 100 -o " + os.path.join(out,"hichipper_peaks.bed"))
			os.system("mv " + os.path.join(out,"hichipper_peaks.bed") + " " + os.path.join(out,"allSamples_temporary_hichipperPeaks.bed"))
			peakfilespersample = [out + '/allSamples_temporary_hichipperPeaks.bed'] * len(samples)
		else:
			peakfilespersample = [out + '/allSamples_temporary_peaks.narrowPeak'] * len(samples)
	
	# Call peaks from within each sample using self-ligation reads
	elif(peaks == "EACH,ALL"):
		click.echo(gettime() + "Calling peaks from all reads for each sample", logf)
		for sample in samples:
		    os.system('cat ' + hicprooutput + '/hic_results/data/' + sample + "/*Pairs >> " + out + "/" + sample +".all.Pairs.tmp")
		    os.system('''awk -v RL=''' + str(halfLength)+r''' '{print $2"\t"$3-RL"\t"$3+RL}' '''+ out + '''/''' + sample +'''.all.Pairs.tmp | awk '$2 > 0 {print $0}' >  ''' + out + '''/''' + sample +'''.all.Pairs.bed.tmp''')
		    os.system('''awk -v RL=''' + str(halfLength)+r''' '{print $5"\t"$6-RL"\t"$6+RL}' '''+ out + '''/''' + sample +'''.all.Pairs.tmp | awk '$2 > 0 {print $0}' >> ''' + out + '''/''' + sample +'''.all.Pairs.bed.tmp''')
		    macs2cmd = 'macs2 callpeak -t ' + out + '/' + sample + '.all.Pairs.bed.tmp ' + macs2_string + " -g " + macs2_genome + " -B -f BED --verbose 0 -n " + out + '/' + sample + "_temporary"
		    click.echo(gettime() +  "macs2 command: " + macs2cmd, logf)
		    os.system(macs2cmd)
		    if not os.path.isfile(out + '/' + sample + '_temporary_peaks.narrowPeak'):
				sys.exit(macs2Error)

		    # Run hichipper modified background correction for each sample using self-ligation reads
		    if (not skip_background_correction):
				click.echo(gettime() + "Performing hichipper-modified background peak calling")
				os.system('Rscript ' + os.path.join(script_dir, 'lambdaProcess.R') + " " + resfrags + " " + out + '/' + sample + "_temporary_treat_pileup.bdg " + out + '/' + sample + "_temporary_control_lambda.bdg " + out)
				click.echo(gettime() + "Modified background signal; performing peak calling")
				click.echo(gettime() +  "MACS2 TEXT OUTPUT INCOMING")
				os.system("macs2 bdgcmp -t " + os.path.join(out,"adjustedTreatment.bdg.tmp") + " -c " + os.path.join(out, "adjustedBackground.bdg.tmp") + " -m qpois -o " + os.path.join(out,"hichipper_qvalue.bdg.tmp"))
				os.system("macs2 bdgpeakcall -i " + os.path.join(out,"hichipper_qvalue.bdg.tmp") + " -c 2 -l 147 -g 100 -o " + os.path.join(out,"hichipper_peaks.bed"))
				os.system("mv " + os.path.join(out,"hichipper_peaks.bed") + " " + out + "/" + sample + "_temporary_hichipperPeaks.bed")
				peakfilespersample.append(out + '/' + sample + '_temporary_hichipperPeaks.bed')
		    else:
				peakfilespersample.append(out + '/' + sample + '_temporary_peaks.narrowPeak')
		    
	
	# Check to see if it's even a valid file
	elif not os.path.isfile(peaks):
		sys.exit('ERROR: Could not identify the ' + peaks + ' file; correctly specify file location or use special variable {COMBINED,EACH},{ALL,SELF} for peaks in .yaml')
	
	# Peaks are apparently determined or user-supplied...
	else:
		click.echo(gettime() + "Using user-defined peaks file " + peaks +  " for analysis.", logf)
		os.system("cp " + peaks + " " + out + "/userSuppliedPeaks.bed.tmp")
		peakfilespersample = [out + "/userSuppliedPeaks.bed.tmp"] * len(samples)
	
	# Do the initial peak padding
	if(int(peak_pad) > 0):
		for peakFile in list(set(peakfilespersample)):
			os.system('''awk -v RL=''' + peak_pad + r''' '{print $1"\t"$2-RL"\t"$3+RL}' '''+ peakFile + ''' | awk '$2 > 0 {print $0}' >  ''' + peakFile + '''_pad.bed.tmp''')
		peakfilespersample = [peakFile + "_pad.bed.tmp" for peakFile in peakfilespersample]
	else:
		click.echo(gettime() + "No initial padding as --peak-pad set to 0 or less", logf)
		
	# Restriction fragment aware padding	
	if skip_resfrag_pad:
		click.echo(gettime() + "Skipping restriction fragment-aware padding based on user flag", logf)
	else:
		click.echo(gettime() + "Performing restriction fragment-aware padding", logf)
		for peakFile in list(set(peakfilespersample)):
			call(['Rscript', os.path.join(script_dir, 'resFragAnchorProcess.R'), resfrags, peakFile])
		peakfilespersample = [peakFile + "rf.tmp" for peakFile in peakfilespersample]
 	logf.close()
 	
 	# Check for UCSC Specification
 	if make_ucsc:
 		ucscoutput = "true"
 	else:
 		ucscoutput = "false"
 	
 	
 	# Call putative interactions
	for i in range(len(samples)):
		hichipperRun = os.path.join(script_dir, 'interactionsCall.sh')	
		cmd = ['bash', hichipperRun, cwd, out, hicprooutput, samples[i], peakfilespersample[i], min_dist, max_dist, merge_gap, str(halfLength), ucscoutput]        
		call(cmd)
		if not os.path.isfile(out + "/" + samples[i] + ".stat"):
			sys.exit('ERROR: something failed at the individual sample level; check the .log file for more info')
	
	# Back to Python
	logf = open(out + "/" + out + ".hichipper.log", 'a')
		
	# QC Report			
	if skip_qc:
		click.echo(gettime() + "Skipping QC report generation since --skip-qc was specified", logf)
	else:
		click.echo(gettime() + "Creating QC report", logf)
		cftg = ' '.join(samples)
		call(['cp', os.path.join(script_dir, 'qcReport_make.Rmd'), out + '/qcReport_make.Rmd'])
		call(['cp', os.path.join(script_dir, 'qcReport.R'), out + '/qcReport.R'])
		cmd = ['Rscript', out + '/qcReport.R', script_dir, str(out), cwd, __version__ , cftg] 		
		click.echo(gettime())
		click.echo(cmd)
		call(cmd)
		
	# diffloop work
	if skip_diffloop:
		click.echo(gettime() + "Skipping diffloop analyses since --skip-diffloop was specified", logf)
	else:
		click.echo(gettime() + "Creating .rds and .mango files", logf)
		cftg = ' '.join(samples)
		cmd = ['Rscript', os.path.join(script_dir, 'diffloop_work.R'), cwd, out, cftg] 
		call(cmd)	
	
	# Temporary File Management
	if keep_temp_files:
		click.echo(gettime() + "Temporary files not deleted since --keep-temp-files was specified", logf)
	else:
		click.echo(gettime() + "Deleting temporary files", logf)
		files = os.listdir(outfolder)
		for file in files:
			if file.endswith(".tmp"):
				os.remove(os.path.join(outfolder,file))
			elif "_temporary_" in file:
				os.remove(os.path.join(outfolder,file))
			elif "qcReport" in file:
				os.remove(os.path.join(outfolder,file))
	click.echo(gettime() + "Done", logf)
	
	logf.close()